File: solutions/0202_happy_number/javascript/solution_fast-slow-pointers.js
/**
 * @param {number} num
 * @return {boolean}
 */
const isHappy = (num) => {
  let slow = sumSquare(num);
  let fast = sumSquare(sumSquare(num));

  while (slow !== fast) {
    slow = sumSquare(slow);
    fast = sumSquare(sumSquare(fast));
  }

  return slow === 1;
};

const sumSquare = (num) => {
  let sum = 0;

  while (num) {
    sum = sum + Math.pow(num % 10, 2);
    num = Math.floor(num / 10);
  }

  return sum;
};

File: solutions/0202_happy_number/java/solution_fast-slow-pointers.java
class Solution {
    public boolean isHappy(int num) {
        int slow = sumSquare(num);
        int fast = sumSquare(sumSquare(num));

        while (slow != fast) {
            slow = sumSquare(slow);
            fast = sumSquare(sumSquare(fast));
        }

        return slow == 1;
    }

    private int sumSquare(int n) {
        int sum = 0;

        while (n > 0) {
            sum += Math.pow(n % 10, 2);
            n /= 10;
        }
        return sum;
    }
}
